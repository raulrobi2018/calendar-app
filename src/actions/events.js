import Swal from "sweetalert2";
import {fetchWithToken} from "../helpers/fetch";
import {formatEvents} from "../helpers/formatEvents";
import {types} from "../types/types";

export const eventStartAddNew = (event) => {
    // obtengo el state usando el getState
    return async (dispatch, getState) => {
        try {
            const {uid, name} = getState().auth;
            const resp = await fetchWithToken("events", event, "POST");
            const body = await resp.json();

            if (body.ok) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                console.log(event);
                dispatch(eventAddNew(event));
            }
        } catch (error) {
            Swal.fire("", "Ha ocurrido un error", "error");
        }
    };
};

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

export const eventUpdate = (event) => ({
    type: types.eventUpdate,
    payload: event
});

export const eventDelete = (event) => ({
    type: types.eventDelete,
    payload: event
});

export const eventsLoad = () => {
    return async (dispatch) => {
        try {
            const resp = await fetchWithToken("events");
            const body = await resp.json();

            //Se transforman las fechas a un formato de Date porque mongo
            //devuelve un string y el Calendar falla si las quiere mostrar
            const events = formatEvents(body.events);

            if (body.ok) {
                dispatch(eventLoaded(events));
            }
        } catch (error) {
            Swal.fire("", "Ha ocurrido un error", "error");
        }
    };
};

const eventLoaded = (events) => ({
    type: types.eventsLoaded,
    payload: events
});
