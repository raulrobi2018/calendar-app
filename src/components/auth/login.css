
@import '../../styles/base.css';

/* Ant Design */
.login-form {
  width: 80%;
  /* max-width: 300px; */
  padding: 50px;
}
.login-form-forgot {
  float: right;
}
.ant-col-rtl .login-form-forgot {
  float: left;
}
.login-form-button {
  width: 100%;
}

.ant-layout-header {
  background-color: blueviolet;
  color: white;
}