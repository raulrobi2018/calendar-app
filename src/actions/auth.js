import { fetchWithOutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import Swal from "sweetalert2";
import { clearEventsOnLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithOutToken("auth", { email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      //Se configuró en 2 horas
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithOutToken(
      "auth/new",
      { name, email, password },
      "POST"
    );

    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      //Se configuró en 2 horas
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchWithToken("auth/renew");

    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      //Se configuró en 2 horas
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
  return (dispatch) => {
    //
    localStorage.clear();
    dispatch(logout());
    dispatch(clearEventsOnLogout());
  };
};

const logout = () => ({ type: types.authLogout });
