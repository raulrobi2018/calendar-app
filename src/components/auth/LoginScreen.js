import React from "react";
import {useDispatch} from "react-redux";
import {startLogin, startRegister} from "../../actions/auth";

import "./login.css";
import {useForm} from "../../hooks/useForm";
import Swal from "sweetalert2";

export const LoginScreen = () => {
    // Example starter JavaScript for disabling form submissions if there are invalid fields
    (function () {
        "use strict";

        // Fetch all the forms we want to apply custom Bootstrap validation styles to
        var forms = document.querySelectorAll(".needs-validation");

        // Loop over them and prevent submission
        Array.prototype.slice.call(forms).forEach(function (form) {
            form.addEventListener(
                "submit",
                function (event) {
                    if (!form.checkValidity()) {
                        event.preventDefault();
                        event.stopPropagation();
                    }

                    form.classList.add("was-validated");
                },
                false
            );
        });
    })();

    const dispatch = useDispatch();

    const [formLoginValues, handleLoginInputChange] = useForm({
        lEmail: "",
        lPassword: ""
    });

    const [formRegisterValues, handleRegisterInputChange] = useForm({
        rName: "",
        rEmail: "",
        rPassword1: "",
        rPassword2: ""
    });

    const {lEmail, lPassword} = formLoginValues;
    const {rName, rEmail, rPassword1, rPassword2} = formRegisterValues;

    const handleLogin = (e) => {
        e.preventDefault();

        dispatch(startLogin(lEmail, lPassword));
    };

    const handleRegister = (e) => {
        e.preventDefault();

        if (rPassword1 !== rPassword2) {
            return Swal.fire(
                "Error",
                "Las contrase単as deben ser iguales",
                "error"
            );
        }

        dispatch(startRegister(rName, rEmail, rPassword1));
    };

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form
                        className="needs-validation"
                        onSubmit={handleLogin}
                        noValidate
                    >
                        <div className="input-group has-validation">
                            <span
                                className="input-group-text"
                                id="inputGroupPrepend"
                            >
                                @
                            </span>
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                name="lEmail"
                                value={lEmail}
                                onChange={handleLoginInputChange}
                                required="required"
                            />
                            <div className="invalid-feedback">
                                Ingrese el nombre de usuario
                            </div>
                        </div>
                        <div className="form-group mt-2">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contrase単a"
                                name="lPassword"
                                value={lPassword}
                                onChange={handleLoginInputChange}
                                required="required"
                                minLength="10"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form
                        onSubmit={handleRegister}
                        className="needs-validation"
                        noValidate
                    >
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name="rName"
                                value={rName}
                                onChange={handleRegisterInputChange}
                                required="required"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                name="rEmail"
                                value={rEmail}
                                onChange={handleRegisterInputChange}
                                required="required"
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contrase単a"
                                name="rPassword1"
                                value={rPassword1}
                                onChange={handleRegisterInputChange}
                                required="required"
                                min="10"
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contrase単a"
                                name="rPassword2"
                                value={rPassword2}
                                onChange={handleRegisterInputChange}
                                required="required"
                                min="10"
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Crear cuenta"
                            />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
};
