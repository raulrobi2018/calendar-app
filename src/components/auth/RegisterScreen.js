import React from "react";
import {useDispatch} from "react-redux";

import {startRegister} from "../../actions/auth";

import {Form, Input, Button, Checkbox, Row, Col, Layout} from "antd";
import {LockOutlined, UserOutlined} from "@ant-design/icons";

import "./login.css";
import "animate.css";
import {Link} from "react-router-dom";

export const RegisterScreen = () => {
    const {Header, Footer, Sider, Content} = Layout;
    const dispatch = useDispatch();

    const onFinish = (values) => {
        console.log("Success:", values);
        const {username, password} = values;
        // dispatch(startLogin(username, password));
    };

    const onFinishFailed = (errorInfo) => {
        console.log("Failed:", errorInfo);
    };

    return (
        <>
            <Layout>
                <Header>MyCalendar</Header>
                <Content>
                    <Row type="flex" justify="space-around" align="middle">
                        <Col xs={2} sm={4} md={3} lg={8} xl={7}></Col>
                        <Col xs={20} sm={16} md={18} lg={8} xl={10}>
                            <Form
                                name="formRegister"
                                initialValues={{
                                    remember: true
                                }}
                                onFinish={onFinish}
                                onFinishFailed={onFinishFailed}
                                className="login-form animate__animated animate__fadeIn"
                            >
                                <Form.Item
                                    name="username"
                                    rules={[
                                        {
                                            required: true,
                                            message: "Ingrese el email"
                                        }
                                    ]}
                                >
                                    <Input
                                        prefix={
                                            <UserOutlined className="site-form-item-icon" />
                                        }
                                    />
                                </Form.Item>

                                <Form.Item
                                    name="password1"
                                    rules={[
                                        {
                                            required: true,
                                            message: "Ingrese la contraseña"
                                        }
                                    ]}
                                >
                                    <Input.Password
                                        prefix={
                                            <LockOutlined className="site-form-item-icon" />
                                        }
                                    />
                                </Form.Item>

                                <Form.Item
                                    name="password2"
                                    rules={[
                                        {
                                            required: true,
                                            message: "Repita la contraseña"
                                        }
                                    ]}
                                >
                                    <Input.Password
                                        prefix={
                                            <LockOutlined className="site-form-item-icon" />
                                        }
                                    />
                                </Form.Item>

                                <Form.Item>
                                    <Button
                                        type="primary"
                                        htmlType="submit"
                                        className="login-form-button"
                                    >
                                        Enviar
                                    </Button>
                                </Form.Item>

                                <hr />
                                <Link to="/auth/login" className="link">
                                    Ya estoy registrado
                                </Link>
                            </Form>
                        </Col>
                        <Col xs={2} sm={4} md={3} lg={8} xl={7}></Col>
                    </Row>
                </Content>
                <Footer>Raul Rodriguez - 2021</Footer>
            </Layout>
        </>
    );
};
