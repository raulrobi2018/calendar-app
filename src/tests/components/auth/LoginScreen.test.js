import React from "react";
import {mount, configure} from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import {Provider} from "react-redux";

import {startLogin} from "../../../actions/auth";

//Configuración necesaria para que funcione mount con React 17
configure({adapter: new Adapter()});
import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";
import {LoginScreen} from "../../../components/auth/LoginScreen";

jest.mock("../../../actions/auth", () => ({
    startLogin: jest.fn()
}));

// Configuracion necesaria para probar dispatch
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);

//Reemplaza la funcionalidad del dispatch del store por una función jest
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);

describe("Testing LoginScreen component", () => {
    test("should display correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("should run the dispatch login", () => {
        wrapper.find('input[name="lEmail"]').simulate("change"),
            {
                target: {
                    name: "lEmail",
                    value: "raul@gmail.com"
                }
            };

        wrapper.find('input[name="lPassword"]').simulate("change"),
            {
                target: {
                    name: "lPassword",
                    value: "123456"
                }
            };

        wrapper.find("form").at(0).prop("onSubmit")({preventDefault() {}});

        expect(startLogin).toHaveBeenCalledWith("raul@gmail.com", "123456");
    });
});
