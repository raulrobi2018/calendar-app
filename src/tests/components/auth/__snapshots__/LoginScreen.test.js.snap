// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginScreen component should display correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        className="needs-validation"
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="input-group has-validation"
        >
          <span
            className="input-group-text"
            id="inputGroupPrepend"
          >
            @
          </span>
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Email"
            required="required"
            type="email"
            value=""
          />
          <div
            className="invalid-feedback"
          >
            Ingrese el nombre de usuario
          </div>
        </div>
        <div
          className="form-group mt-2"
        >
          <input
            className="form-control"
            minLength="10"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            required="required"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        className="needs-validation"
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            required="required"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Email"
            required="required"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            min="10"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            required="required"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            min="10"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            required="required"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
